roxie/ccd/ccdserver.cpp

```diff
@@ -2394,13 +2393,14 @@ class RecordPullerThread : public RestartableThread
     {
         if (traceStartStop)
             DBGLOG("RecordPullerThread::stop");
-        {
-            CriticalBlock c(crit); // stop is called on our consumer's thread. We need to take care calling stop for our input to make sure it is not in mid-nextRow etc etc.
-            if (inputStream)
-                inputStream->stop();
-            eof = true;
-        }
+
+        //Force the reading thread to terminate
+        eof = true;
```

Comment:
Not sure this will stop as quickly as calling stop() would - for example it's still going to wait for upstream activities that may have long timeouts...

Reply:
I think the behaviour is very similar - but a fix would make it better.  There are two cases:
i) other thread is not inside nextRow().
This will set eof to true, which means next time nextRow() is called it will immediately return.  The join will then succeed, and stop will then be called.
Previously it would have grabbed the critical section, called stop, and set the flag.  The next time nextRow() is called it will immediately return.  The join will then succeed.   It may be slightly worse performance because the stop() will not overlap with any processing in the other thread.
ii) other thread is inside nextRow()
Now it will set eof, wait for the thread to terminate then call stop().  Previously it would have waited for the row to be processed, then set eof, called stop() and waitied for the other thread to join.  Again stop() processing will not be overlapped with nextRow() processing.

I think the gain from not requiring a critical section every row (whether or not it is asynchronously closed) is likely to outweigh the cost.

The (minor) improvement that would be worth doing (for old and new code) would be to recheck eof after nextRow() is called - and avoid calling processGroup() etc..  (The append is probably fine.)

Reply:
If you are right, then the behaviour in case ii is suboptimal today - things may block for a while.

There may be other mechanisms (eg checkAbort) that make it ok in most cases

Reply:
On the other hand if the reading thread has finished the new code will be faster (since there will be no need for the critical section).